src/anndata/_core/anndata.py

```diff
@@ -85,6 +85,40 @@ def _check_2d_shape(X):
         raise ValueError(msg)
 
 
+def _infer_shape_for_axis(
```

Comment:
2 things about these functions.

1. Do they belong in `utils`? I see we have a bunch at the top of the file here so not sure
2. Should we do some sort of validation that all the shapes match? I think this happens anyway elsewhere in the codebase e.g., https://github.com/scverse/anndata/blob/main/src/anndata/_core/aligned_mapping.py#L70-L100

Reply:
1. I think the logic for where things are is mostly “the deepest module level shared by all spots using it”, so something used all throughout would be in `src/anndata/_utils.py`, something only used by IO routines would be in `src/anndata/io/_utils.py` and something only used in one spot would be in the same file.
2. We allow inferring the shape from whatever is specified right? So figuring out the shape is no longer strictly before setting aligned mappings.

   Technically the optimal way to do things would then be to kinda coerce everything together, infer the shape, then at the end set everything. But I think it’s fine to do things a bit less elegantly if that saves work (the result is the same: An error if the shape can’t be unambiguously inferred)